eth-rpc/client/clientdispatch.nim

```diff
@@ -77,37 +89,102 @@ proc connect*(self: RpcClient, address: string, port: Port) {.async.} =
+import ../ jsonmarshal
+
+proc createRpcProc(procName, parameters, callBody: NimNode): NimNode =
+  # parameters come as a tree
+  var paramList = newSeq[NimNode]()
+  for p in parameters: paramList.add(p)
+
+  result = newProc(procName, paramList, callBody)           # build proc
+  result.addPragma ident"async"                             # make proc async               
+  result[0] = nnkPostFix.newTree(ident"*", newIdentNode($procName))  # export this proc
+
+proc toJsonArray(parameters: NimNode): NimNode =
+  # outputs an array of jsonified parameters
+  # ie; %[%a, %b, %c]
+  parameters.expectKind nnkFormalParams
+  var items = newNimNode(nnkBracket)
+  for i in 2 ..< parameters.len:
+    let curParam = parameters[i][0]
+    if curParam.kind != nnkEmpty:
+      items.add(nnkPrefix.newTree(ident"%", curParam))
+  result = nnkPrefix.newTree(newIdentNode("%"), items)
+
+proc createRpcFromSig*(rpcDecl: NimNode): NimNode =
+  # Each input parameter in the rpc signature is converted
+  # to json with `%`.
+  # Return types are then converted back to native Nim types.
+  let iJsonNode = newIdentNode("JsonNode")
+
+  var parameters = rpcDecl.findChild(it.kind == nnkFormalParams).copy
+  # ensure we have at least space for a return parameter  
+  if parameters.isNil or parameters.kind == nnkEmpty or parameters.len == 0:
+    parameters = nnkFormalParams.newTree(iJsonNode)
+
+  let
+    procName = rpcDecl.name
+    pathStr = $procName
+    returnType =
+      # if no return type specified, defaults to JsonNode
+      if parameters[0].kind == nnkEmpty: iJsonNode
+      else: parameters[0]
+    customReturnType = returnType != iJsonNode
+
+  # insert rpc client as first parameter
+  parameters.insert(1, nnkIdentDefs.newTree(ident"client", ident"RpcClient", newEmptyNode()))
+
+  let
+    jsonParamIdent = genSym(nskVar, "jsonParam")  # variable used to send json to the server
+    jsonParamArray = parameters.toJsonArray()     # json array of marshalled parameters
+  var
+    # populate json params - even rpcs with no parameters have an empty json array node sent
+    callBody = newStmtList().add(quote do:
+      var `jsonParamIdent` = `jsonParamArray`
+    )
+
+  # convert return type to Future
+  parameters[0] = nnkBracketExpr.newTree(ident"Future", returnType)
+  # create rpc proc
+  result = createRpcProc(procName, parameters, callBody)
+
+  let
+    rpcResult = genSym(nskLet, "res") # temporary variable to hold `Response` from rpc call
+    procRes = ident"result"           # proc return variable
+    jsonRpcResult =                   # actual return value, `rpcResult`.result
+      nnkDotExpr.newTree(rpcResult, newIdentNode("result"))
+  
+  # perform rpc call
+  callBody.add(quote do:
+    let `rpcResult` = await client.call(`pathStr`, `jsonParamIdent`)          # `rpcResult` is of type `Response`
+    if `rpcResult`.error: raise newException(ValueError, $`rpcResult`.result) # TODO: is raise suitable here? 
```

Comment:
I think an exception here is appropriate, but you should define your own exception types instead of using `ValueError` everywhere.

The exception will be put in the `Future` object representing the call. There might be a short-cut for doing this without raising an exception. If there isn't such a shortcut now, @cheatfate can add one to the new `asyncdispatch2` lib.